jax/scipy/spatial/distance.py

```diff
@@ -0,0 +1,150 @@
+import jax.numpy as jnp
+from jax import jit, vmap
+
+# define primitives for distance functions according to scipy.spatial.distance
+# which can be batched over the first dimension and then used inside jit and vmap
+# as well as used for implementing cdist and pdist versions
+
+primitives = {
+    "braycurtis": lambda x, y: jnp.sum(jnp.abs(x - y), axis=-1)
+    / jnp.sum(jnp.abs(x + y), axis=-1),
+    "canberra": lambda x, y: jnp.sum(
+        jnp.abs(x - y) / (jnp.abs(x) + jnp.abs(y)), axis=-1
+    ),
+    "chebyshev": lambda x, y: jnp.max(jnp.abs(x - y), axis=-1),
+    "cityblock": lambda x, y: jnp.sum(jnp.abs(x - y), axis=-1),
+    "correlation": lambda x, y: 1
+    - jnp.sum((x - y) ** 2, axis=-1)
+    / ((jnp.sum(x**2, axis=-1) * jnp.sum(y**2, axis=-1)) ** 0.5),
+    "cosine": lambda x, y: 1
+    - jnp.sum(x * y, axis=-1)
+    / (jnp.linalg.norm(x, axis=-1) * jnp.linalg.norm(y, axis=-1)),
+    "dice": lambda x, y: jnp.sum(x != y, axis=-1)
+    / (jnp.sum(x != 0, axis=-1) + jnp.sum(y != 0, axis=-1)),
+    "euclidean": jnp.linalg.norm,
+    "hamming": lambda x, y: jnp.sum(x != y, axis=-1) / x.shape[-1],
+    "jaccard": lambda x, y: jnp.sum(x != y, axis=-1)
+    / jnp.sum((x != y) | (x != 0) | (y != 0), axis=-1),
+    "kulczynski1": lambda x, y: (jnp.sum(x != y, axis=-1) - jnp.sum(x == y, axis=-1))
+    / (jnp.sum(x != y, axis=-1) + jnp.sum(x == y, axis=-1)),
+    "mahalanobis": lambda x, y, V: jnp.sqrt(jnp.sum(((x - y) / V) ** 2, axis=-1)),
+    "matching": lambda x, y: jnp.sum(x != y, axis=-1) / x.shape[-1],
+    "minkowski": lambda x, y, p: jnp.sum(jnp.abs(x - y) ** p, axis=-1) ** (1 / p),
+    "rogerstanimoto": lambda x, y: (jnp.sum(x != y, axis=-1) - jnp.sum(x == y, axis=-1))
+    / (2 * jnp.sum(x != y, axis=-1) + jnp.sum(x == y, axis=-1)),
+    "russellrao": lambda x, y: jnp.sum(x != y, axis=-1) / x.shape[-1],
+    "seuclidean": lambda x, y, V: jnp.sqrt(jnp.sum(((x - y) / V) ** 2, axis=-1)),
+    "sokalmichener": lambda x, y: (jnp.sum(x != y, axis=-1) - jnp.sum(x == y, axis=-1))
+    / (jnp.sum(x != y, axis=-1) + jnp.sum(x == y, axis=-1)),
+    "sokalsneath": lambda x, y: (jnp.sum(x != y, axis=-1) - jnp.sum(x == y, axis=-1))
+    / (jnp.sum(x != y, axis=-1) + jnp.sum(x == y, axis=-1)),
+    "sqeuclidean": lambda x, y: jnp.sum((x - y) ** 2, axis=-1),
+    "yule": lambda x, y: (jnp.sum(x != y, axis=-1) - jnp.sum(x == y, axis=-1))
+    / (jnp.sum(x != y, axis=-1) + jnp.sum(x == y, axis=-1)),
+}
+
+# vectorize primitives for batched, jit'ed evaluation. Take special care for the primitives that
+# take additional arguments (p, V)
+_METRICS = {}
+for k, v in primitives.items():
+    if k in ("minkowski", "mahalanobis", "seuclidean"):
+        _METRICS[k] = jit(vmap(v, in_axes=(0, 0, None)))
+    else:
+        _METRICS[k] = jit(vmap(v))
+
+
+def pdist(X, metric="euclidean", p=2, V=None):
```

Comment:
The scipy function allows general `**kwargs` to be forwarded to the distance metric – that would prevent the need for special cases below.